### PR TITLE
(feat) Impl SegmentList to replace ArrayDequeue in LogManagerImpl, #335

### LINKED ISSUE #335: logsInMemory needs a more appropriate deque
### Your question

LogManagerImpl#logsInMemory needs a more appropriate deque, now the `remove range` operation cause data movement

### Environment

- SOFAJRaft version:
- JVM version (e.g. `java -version`):
- OS version (e.g. `uname -a`):
- Maven version:
- IDE version:

### PR DESCRIPTION
**Pull request by killme2008**
### Motivation:

See #335 

### Modification:

* Introduce a new data structure `SegmentList`  that is a list of segment, every segment is an array contains 128 elements, and only supports removing elements from start or end:

```
                            segments
       /                        |                       \
segment                      segment                      segment 
[0, 1 ...  127]             [128, 129 ... 255]          [256, 1 ... 383] 
```

The `removeFromFirstWhen` and `removeFromLastWhen` methods accept a predicate and delete the elements from start or end until the predicate returns false. The ` remove` operation only set the elements to be null and remeber the deleted poistion,do not have any array elements movement at all. If a segment is empty after `remove`, it will  be removed from segment list too.
Also, I use the `Recyclable` to reuse the `Segment` instances.

The `SegmentList` performance in benchmark is alaways better than `ArrayDequeue` in `SegmentListTest#simpleBenchmark`, at most 10% improvement(The benchmark has some random factor). 

* Replace `ArrayDeque` in `LogManagerImpl` with `SegmentList`
* Fixed deadlock in `RepeatedTimer#destroy` and `RepeatedTimer#run`

### Result:

Fixes #335

### CODE REVIEW

File: jraft-core/src/main/java/com/alipay/sofa/jraft/util/SegmentList.java [hunk 1]
```diff
@@ -0,0 +1,347 @@
...
+/**
+ * A list implementation based on segments.Only supports removing elements from start or end.
+ * The list keep the elements in a segment list, every segment contains at most 128 elements.
+ *
+ * @author boyan(boyan@antfin.com)
+ *
+ */
+public class SegmentList<T> {
+    private static final int             SEGMENT_SHIFT = 8;
+    public static final int              SEGMENT_SIZE  = 2 << (SEGMENT_SHIFT - 1);
```
Comment (OpenOpened) [inline jraft-core/src/main/java/com/alipay/sofa/jraft/util/SegmentList.java]: every segment contains at most 128 elements, but **SEGMENT_SIZE 2 << 7 equals 256**.

Comment (killme2008) [reply jraft-core/src/main/java/com/alipay/sofa/jraft/util/SegmentList.java]: Fixed.  128 is a balance between time and space.

File: jraft-core/src/main/java/com/alipay/sofa/jraft/util/SegmentList.java [hunk 2]
```diff
@@ -0,0 +1,348 @@
...
+                this.firstOffset = 0;
+            }
+        } else {
+            this.firstOffset = this.size = 0;
+        }
+    }
+
+    @SuppressWarnings("unchecked")
+    public void addAll(final Collection<T> coll) {
+        Object[] src = coll.toArray();
```
Comment (fengjiachun) [inline jraft-core/src/main/java/com/alipay/sofa/jraft/util/SegmentList.java]: 入参为 Collection 类型更通用，原有 ArrayDeque 的设计也没有避免一次 toArray 的多余 memcopy

但我还是建议这里针对 ArrayList 做一个 fast path 优化，在入参 coll 为 ArrayList 类型且 `UnsafeUtil.hasUnsafe() == true` 的情况下通过 `UnsafeReferenceFieldUpdater` 直接拿到 ArrayList 中的 `object[]`，用于后面的 copy，再这样节省了一次 memcopy；

`AppendEntries()` 是 jraft 中的热点执行路径，会调用  `addAll ()` 来 append log，所以我认为即使代码丑一点，但是优化是值得的，理论上会减少大量的 toArray 带来的临时 'object[]' 的创建和 memcopy

**Comment by fengjiachun** [comment]
```
                            segments
       /                        |                       \
segment                      segment                      segment 
[0, 1 ...  127]             [128, 129 ... 255]          [256, 1 ... 383] 
```

这个图是不是可以加到 `SegmentList` 的类注释里面

File: jraft-core/src/main/java/com/alipay/sofa/jraft/util/SegmentList.java [hunk 3]
```diff
@@ -0,0 +1,348 @@
...
+                RecycleUtil.recycle(this.segments.pollFirst());
+                firstSeg = getFirst();
+                this.firstOffset = 0;
+            }
+        }
+    }
+
+    public void clear() {
+        while (!this.segments.isEmpty()) {
+            RecycleUtil.recycle(this.segments.pollLast());
```
Comment (fengjiachun) [inline jraft-core/src/main/java/com/alipay/sofa/jraft/util/SegmentList.java]: 这里貌似存在内存泄漏，对于每一个 segment， 应该将内部数组维持的所有引用全部置 null 再 recycle ？
否则一直会维持强引用，相应元素得不到释放

Comment (killme2008) [reply jraft-core/src/main/java/com/alipay/sofa/jraft/util/SegmentList.java]: 不会的，最终调用到 `Segment#recycle` 方法，里面会填充null

Comment (killme2008) [reply jraft-core/src/main/java/com/alipay/sofa/jraft/util/SegmentList.java]: `Segment#recycle` => `Segment#clear` => `Arrays.fill(this.elements, null);`

File: jraft-core/src/main/java/com/alipay/sofa/jraft/util/SegmentList.java (hunk 2 above)
Comment (killme2008) [reply jraft-core/src/main/java/com/alipay/sofa/jraft/util/SegmentList.java]: 嗯，搞了下，测试下来，稳定提升在 15% 以上 👍 ，不过需要将 UnsafeReferenceFieldUpdater 从 rheakv 拷贝一份到 jraft-core 。不做 move，因为不确定外面有没有人依赖了。

**Comment by killme2008** [comment]
Updated:

1. `SegmentList` constructor accepts `boolean recycleSegment` to choose whether to enable recycling segment.
2. `SegmentList#addAll` adds a fast path for `ArrayList`, suggested by @fengjiachun 
3. Replace `pendingMetaQueue` in `BallotBox` with `SegmentList` too.
4. Move some utilities from rheakv-core to jraft-core, such as `ReferenceFieldUpdater` etc.

File: jraft-core/src/main/java/com/alipay/sofa/jraft/util/SegmentList.java (hunk 3 above)
Comment (fengjiachun) [reply jraft-core/src/main/java/com/alipay/sofa/jraft/util/SegmentList.java]: 恩，没注意到 clear

File: jraft-core/src/main/java/com/alipay/sofa/jraft/util/SegmentList.java [hunk 4]
```diff
@@ -0,0 +1,391 @@
...
+            lastSeg.addAll(src, srcPos, len);
+            srcPos += len;
+            this.size += len;
+        }
+    }
+
+    private Object[] coll2Array(final Collection<T> coll) {
+        Object[] src = null;
+        if (coll instanceof ArrayList) {
+            if (UnsafeUtil.hasUnsafe()) {
```
Comment (fengjiachun) [inline jraft-core/src/main/java/com/alipay/sofa/jraft/util/SegmentList.java]: `if (coll instanceof ArrayList && UnsafeUtil.hasUnsafe())`
否则在没有 unsafe 的平台运行  src 会没有被赋值

**Pull request merged**